File: windows/WinLock.Service/Bluetooth/ProximityState.cs
namespace WinLock.Service.Bluetooth;

/// <summary>
/// Bluetooth proximity state. BLE presence is a proximity signal only and never
/// authorizes operations. The spec's combined "AUTHENTICATED" state is derived
/// client-side (nearby + a valid authenticated session), not reported here.
/// </summary>
public enum ProximityState
{
    Unknown = 0,
    Nearby = 1,
    Away = 2,
}

/// <summary>A snapshot of Bluetooth proximity for the monitored devices.</summary>
public sealed record ProximitySnapshot(
    ProximityState State,
    string? DeviceId,
    int? Rssi,
    DateTimeOffset UpdatedAtUtc)
{
    public static ProximitySnapshot Unknown() =>
        new(ProximityState.Unknown, null, null, DateTimeOffset.UtcNow);
}

File: iphone/src/bluetooth/proximityState.ts
import type { ProximityStateString } from '../types/protocol';

/**
 * Combined proximity state. The server reports UNKNOWN / NEARBY / AWAY; the
 * spec's AUTHENTICATED state is derived client-side (nearby + a valid
 * authenticated session). BLE presence never authenticates anything.
 */
export function deriveProximity(
  serverState: ProximityStateString,
  isAuthenticated: boolean,
): ProximityStateString {
  if (isAuthenticated && serverState === 'NEARBY') {
    return 'AUTHENTICATED';
  }
  return serverState;
}